pkg/controller/manager/vlanconfig/controller.go

```diff
@@ -83,16 +83,49 @@ func (h Handler) SetClusterNetworkUnready(_ string, vs *networkv1.VlanStatus) (*
 	return vs, nil
 }
 
-func (h Handler) ensureClusterNetwork(name string) error {
-	if _, err := h.cnCache.Get(name); err != nil && !apierrors.IsNotFound(err) {
+func (h Handler) ensureClusterNetwork(vc *networkv1.VlanConfig) error {
+	name := vc.Spec.ClusterNetwork
+	curCn, err := h.cnCache.Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
 		return err
-	} else if err == nil {
+	}
+
+	MTU := utils.DefaultMTU
+	if utils.IsValidMTU(vc.Spec.Uplink.LinkAttrs.MTU) && vc.Spec.Uplink.LinkAttrs.MTU != 0 {
+		MTU = vc.Spec.Uplink.LinkAttrs.MTU
+	}
+	targetMTU := fmt.Sprintf("%v", MTU)
+
+	// check if the configured VC MTU value is updated to ClusterNetwork annotations
+	if curCn != nil {
+		curMTU := curCn.Annotations[utils.KeyUplinkMTU]
+		if curMTU == targetMTU {
+			// do not compare KeyMTUSourceVlanConfig, which is only used for reference
+			return nil
+		}
+		// update the new MTU, e.g. a new MTU value is set on the vlanconfig
+		cnCopy := curCn.DeepCopy()
+		if cnCopy.Annotations == nil {
+			cnCopy.Annotations = make(map[string]string, 2)
+		}
+		cnCopy.Annotations[utils.KeyUplinkMTU] = targetMTU
+		cnCopy.Annotations[utils.KeyMTUSourceVlanConfig] = vc.Name
+		if _, err := h.cnClient.Update(cnCopy); err != nil {
+			return fmt.Errorf("failed to update cluster network %s annotation %s with MTU %s error %w", name, utils.KeyUplinkMTU, targetMTU, err)
+		}
+		klog.Infof("update cluster network %s annotation %s to %s", name, utils.KeyUplinkMTU, targetMTU)
 		return nil
 	}
 
 	// if cn is not existing
 	cn := &networkv1.ClusterNetwork{
```

Comment:
would there be a scenario where clusternetwork does not exist before a vlanconfig is created?

Reply:
Generally, vlanconfig should not come before clusternetwork, it is cross-ensured by this PR

cn onDelete:
https://github.com/harvester/network-controller-harvester/pull/149/files#diff-702e1e8c8be9e7fd0f87527272d0fff1879b6aeb663d616640d4dcd0d17238e4R96

vlanconfig onCreate
https://github.com/harvester/network-controller-harvester/pull/149/files#diff-e30ba60a51ad3f7ae0d7a37a760cccb6c14c47fa0fa07a5b76fc1b62bd5d0112R72

If unluckily the clusternetwork is not existing (e.g. hackly removed the webhook to create a dangling vlanconfig), then L121 will create it.
https://github.com/harvester/network-controller-harvester/pull/149/files#diff-73eb21b655ff1d2a6f308f3f893c6c248f31d67f90a5d9bab4029c61e2edea31R100